rwkv.cpp

```diff
@@ -84,72 +129,241 @@ inline enum rwkv_error_flags operator|=(enum rwkv_error_flags & a, enum rwkv_err
-    RWKV_ASSERT_FALSE_MSG(RWKV_ERROR_FILE_READ, fread(dest, sizeof(int32_t), 1, file) == 1, "Failed to read an int32 value from a file (%s)", name);
+// Reads a single uint32 value from a file.
+bool fread_uint32(FILE * file, uint32_t & dest) {
+    return fread((void *) &dest, sizeof(uint32_t), 1, file) == 1;
+}
+
+// Reads a single string value from a file.
+bool fread_string(FILE * file, size_t length, std::string & dest) {
+    dest.resize(length);
+    return fread((void *) dest.data(), length, 1, file) == 1;
+}
+
+// Reads a single data buffer from a file.
+bool fread_data(FILE * file, size_t length, void * dest) {
+    return fread(dest, length, 1, file) == 1;
+}
+
+// Writes a single uint32 value to a file.
+bool fwrite_uint32(FILE * file, const uint32_t value) {
+    return fwrite((const void *) &value, sizeof(uint32_t), 1, file);
+}
+
+// Writes a single string value to a file.
+bool fwrite_string(FILE * file, const std::string & value) {
+    return fwrite((const void *) value.data(), value.length(), 1, file) == 1;
+}
+
+// Writes a single data buffer to a file.
+bool fwrite_data(FILE * file, const void * data, const size_t length) {
+    return fwrite(data, length, 1, file) == 1;
+}
+
+// --- File data structures ---
+
+#define TYPE_UNKNOWN TYPE_COUNT
+
+enum rwkv_type {
+    TYPE_F32,
+    TYPE_F16,
+    TYPE_Q4_0,
+    TYPE_Q4_1,
+    TYPE_Q4_1_O, // Unsupported
+    TYPE_Q4_2, // Unsupported
+    TYPE_Q4_3, // Unsupported
+    TYPE_Q5_0,
+    TYPE_Q5_1,
+    TYPE_Q8_0,
+    TYPE_COUNT
+};
+
+#define GGML_TYPE_UNKNOWN GGML_TYPE_COUNT
+
+static const enum ggml_type type_to_ggml[TYPE_COUNT + 1] = {
+    GGML_TYPE_F32,     /* F32    */
+    GGML_TYPE_F16,     /* F16    */
+    GGML_TYPE_Q4_0,    /* Q4_0   */
+    GGML_TYPE_Q4_1,    /* Q4_1   */
+    GGML_TYPE_UNKNOWN, /* Q4_1_O */
+    GGML_TYPE_UNKNOWN, /* Q4_2   */
+    GGML_TYPE_UNKNOWN, /* Q4_3   */
+    GGML_TYPE_Q5_0,    /* Q5_0   */
+    GGML_TYPE_Q5_1,    /* Q5_1   */
+    GGML_TYPE_Q8_0,    /* Q8_0   */
+    GGML_TYPE_COUNT    /* COUNT  */
+};
+
+static const enum rwkv_type type_from_ggml[GGML_TYPE_COUNT + 1] = {
+    TYPE_F32,    /* F32   */
+    TYPE_F16,    /* F16   */
+    TYPE_Q4_0,   /* Q4_0  */
+    TYPE_Q4_1,   /* Q4_1  */
+    TYPE_Q4_2,   /* Q4_2  */
+    TYPE_Q4_3,   /* Q4_3  */
+    TYPE_Q5_0,   /* Q5_0  */
+    TYPE_Q5_1,   /* Q5_1  */
+    TYPE_Q8_0,   /* Q8_0  */
+    TYPE_COUNT,  /* Q8_1  */
+    TYPE_COUNT,  /* I8    */
+    TYPE_COUNT,  /* I16   */
+    TYPE_COUNT,  /* I32   */
+    TYPE_COUNT,  /* COUNT */
+};
+
+static const char * type_to_string[TYPE_COUNT + 1] = {"float32", "float16", "Q4_0", "Q4_1", "Q4_1_O", "Q4_2", "Q4_3", "Q5_0", "Q5_1", "Q8_0", "unknown"};
+
+enum rwkv_type type_from_string(const char * str) {
+    for (int ord = 0; ord < TYPE_COUNT; ord++)
+        if (strcmp(str, type_to_string[ord]) == 0)
+            return (enum rwkv_type) ord;
+
+    return TYPE_UNKNOWN;
+}
+
+struct file_header {
```

Comment:
```suggestion
struct rwkv_file_header {
```
Not a blocker, but I would prefer to prefix `rwkv.cpp` specific structures and functions with `rwkv_`. Just a style preference

Reply:
sorry, I think I tried this but didn't know rwkv_ is also for internal functions and not just api functions